crategen/converters/tes_converter.py

```diff
@@ -1,52 +1,117 @@
-            "result": [{"@id": output.get("url", ""), "name": output.get("path", "")} for output in outputs],
-            "startTime": convert_to_iso8601(creation_time),
-            "endTime": convert_to_iso8601(end_time),
+            "@id": data_tes.id,
+            "name": data_tes.name,
+            "description": data_tes.description,
+            "instrument": executors[0].image if executors else None,
+            "object": [
+                {"@id": input.url, "name": input.path} for input in data_tes.inputs
+            ],
+            "result": [
+                {"@id": output.url, "name": output.path} for output in data_tes.outputs
+            ],
+            "startTime": data_tes.creation_time,
+            "endTime": end_time,
         }
+
+        validate_wrroc_tes(wrroc_data)
         return wrroc_data
 
-    def convert_from_wrroc(self, wrroc_data):
-        # Validate and extract data with defaults
-        id = wrroc_data.get("@id", "")
-        name = wrroc_data.get("name", "")
-        description = wrroc_data.get("description", "")
-        instrument = wrroc_data.get("instrument", "")
-        object_data = wrroc_data.get("object", [])
-        result_data = wrroc_data.get("result", [])
-        start_time = wrroc_data.get("startTime", "")
-        end_time = wrroc_data.get("endTime", "")
-
-        # Convert from WRROC to TES
-        tes_data = {
-            "id": id,
-            "name": name,
-            "description": description,
-            "executors": [{"image": instrument}],
-            "inputs": [{"url": obj.get("@id", ""), "path": obj.get("name", "")} for obj in object_data],
-            "outputs": [{"url": res.get("@id", ""), "path": res.get("name", "")} for res in result_data],
-            "creation_time": start_time,
-            "logs": [{"end_time": end_time}],  # Added to logs
-        }
-        return tes_data
+    def convert_from_wrroc(self, data: dict) -> dict:
+        """
+        Convert WRROC data to TES format.
+
+        Args:
+            data: The input WRROC data.
+
+        Returns:
+            The converted TES data.
+
+        Raises:
+            ValidationError: If WRROC data is invalid.
+        """
+        # Validate WRROC data
+        try:
+            data_wrroc = WRROCDataTES(**data)
+        except ValidationError as e:
+            raise ValueError(f"Invalid WRROC data: {e.errors()}") from e
+
+        # Convert URL strings to AnyUrl
+        tes_inputs = [
+            TESInput(url=AnyUrl(url=obj.id), path=obj.name) for obj in data_wrroc.object
+        ]
+        tes_outputs = [
+            TESOutput(url=AnyUrl(url=data_wrroc.result.id), path=data_wrroc.result.name)
+        ]
+
+        # Ensure 'image' and 'command' fields are provided
+        tes_executors = [
+            TESExecutor(image=data_wrroc.instrument or "", command=[])
+        ]  # Provide default empty list for command
```

Comment:
I agree with your points. I’ll update the logic to:

    Source the command from the input data (e.g., `data_wrroc`).
    Add an error message if no command or environment is provided, rather than proceeding with an invalid TES task.